### PR TITLE
add WinResized autocmd

### PR DESCRIPTION
**Pull request by sunjon**
Adds WinResized event as per https://github.com/neovim/neovim/issues/11532.

### CODE REVIEW

**Comment by sunjon** [comment]
I don't think this PR works as intended. Triggering the event when the active window is resized is not enough for a plugin to know when to resize other windows.  If an api command or mouse resizes a split that's not adjacent to the active window, no event is generated.

Any advice on how to change this so the event is generated after any window is resized?

File: src/nvim/edit.c [hunk 1]
```diff
@@ -1487,6 +1487,12 @@ static void ins_redraw(
     do_autocmd_winscrolled(curwin);
   }
 
+  // Trigger Resized if viewport changed.
+  if (ready && has_event(EVENT_WINRESIZED)
+      && win_did_resize(curwin)) {
```
Comment (bfredl) [inline src/nvim/edit.c:1492]: may I suggest `FOR_ALL_WINDOWS_IN_TAB(wp, curtab)` instead of `curwin` :D

File: src/nvim/window.c [hunk 2]
```diff
@@ -5006,6 +5006,36 @@ void do_autocmd_winscrolled(win_T *wp)
...
+  tv_dict_add_nr(dict, S_LEN("height_new"), wp->w_height);
+
+  tv_dict_set_keys_readonly(dict);
+  textlock++;
+  apply_autocmds(EVENT_WINRESIZED, NULL, NULL, false, curbuf);
+  textlock--;
+  tv_dict_clear(dict);
+
+  wp->w_last_width = wp->w_width;
+  wp->w_last_height = wp->w_height;
```
Comment (romgrk) [inline src/nvim/window.c:5036]: This is going to cause an use-after-free if the autocmd handler closes `wp`.

**Comment by romgrk** [comment]
To continue a conversation from gitter:

> sunjon: I'm using it pretty much as WinResized, which works - except a) it doesn't trigger on <C-w> <foo> commands, and requires some extra parsing to work out what exactly resized

Note: I'm going to use "events" as meaning "autocommands".

The problem here is precisely the problem that #13247 is trying to solve. The current model for events is that each iteration of the loop for each State checks if events should be dispatched for the current window/buffer. This worked before asynchronicity was introduced but obviously that model doesn't work now that we have the neovim API.
The changes that #13247 brings are that the previous checks are kept, but in addition for the neovim API and where it applies in the rest of the codebase, each function checks for events that might need triggering following what the function did. I'm not going to do that for all events because eg checking for TextChanged everywhere where text can change would be quite inconvenient. But for new events such as WinScrolled I have been adding the checks in individual places in the codebase where it makes sense (eg window split & window close).

Where I'm at with #13247:
 - In general, it works but the fact that autocommands can close stuff causes use-after-free annoying situations (like my previous comment).
 - I might a been trying to bite more than I can chew because I wanted to solve #10070 and #13201 in the same PR but the code changes are probably too big to fit in a single PR even thought the same underlying patterns would be used to solve all those issues.
 - Need to add more tests.

I think finishing #13247 would show the good patterns to use to add further autocommands because some of what is done in this PR is essentially using the previous synchronous patterns which won't work.

I might also add that WinScrolled is more like WinLayoutChanged, and as far as I remember it can already act as a WinResized because it checks for width/height changes. It's that way because changing the size of a window changes its visible scroll area, and that is what plugins listening for that event are looking for, they wanna know when the visible area changes. The doc about that isn't as clear as I remembered it so it should probably be changed. I'm not sure if it makes sense to add a WinResized event in that case even though it's conceptually clearer to have both.

Last thing, there is [this discussion](https://github.com/vim/vim/issues/5181) with Bram to talk about event compatibility for WinScrolled, there were mentions of using `expand('<awinid>')` instead of `v:event` for communicating the window triggering the event, it wasn't settled but it might be good to communicate to ensure compatibility whatever we do.

**Comment by b0o** [comment]
> I might also add that WinScrolled is more like WinLayoutChanged, and as far as I remember it can already act as a WinResized because it checks for width/height changes. It's that way because changing the size of a window changes its visible scroll area, and that is what plugins listening for that event are looking for, they wanna know when the visible area changes. The doc about that isn't as clear as I remembered it so it should probably be changed. I'm not sure if it makes sense to add a WinResized event in that case even though it's conceptually clearer to have both.

Thank you for mentioning this, @romgrk. I came here looking for a `WinResized` event, but knowing this, `WinScrolled` will mostly work for my purposes. I think this should absolutely be documented better, I have spent quite a while searching for a way to trigger on resize. I think `WinLayoutChanged` would actually be a much better name for that event, but I guess that ship has already sailed.

**Pull request closed without merging**